File: LeetCode/Python/solutions/p0022.py
from typing import List


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        def helper(left, right, curr):

            if len(curr) == n * 2:
                ans.append(curr)

            if left < n:
                helper(left + 1, right, curr+ "(")

            if left > right:
                helper(left, right + 1, curr + ")")

        helper(0, 0, "")

        return ans

if __name__ == "__main__":
    solution = Solution()

    n = 3
    print(f"return: {solution.generateParenthesis(n)}")

    n = 1
    print(f"return: {solution.generateParenthesis(n)}")

File: LeetCode/Java/src/solutions/p0022_GenerateParentheses/P0022.java
package solutions.p0022_GenerateParentheses;

import java.util.ArrayList;
import java.util.List;

public class P0022 {
    List<String> ans;

    public List<String> generateParenthesis(int n) {
        /*
         * Solution v1: Brute Force
         *
         * Runtime: 1 ms @ (beats) 92.80%
         * Memory Usage: 42.4 MB @ (beats) 70.10%
         */
        this.ans = new ArrayList<String>();
        builder(0, 0, n, "");
        return this.ans;
    }

    public void builder(int left, int right, int n, String curr) {
        if (curr.length() == n * 2) {
            this.ans.add(curr);
        }

        if (left < n) {
            this.builder(left + 1, right, n, curr.concat("("));
        }

        if (left > right) {
            this.builder(left, right + 1, n, curr.concat(")"));
        }

    }

}
